File: packages/core/src/nn/attention.ts
import { PipelineManager } from "../gpu/pipeline.js";
import { BufferPool } from "../gpu/buffer-pool.js";
import { BitLinear } from "./bitlinear.js";
import { type BindGroupCache, createBGCache, clearBGCache, cachedBG } from "./bg-cache.js";
import type { ModelConfig, KVCache } from "../types.js";
import { headDim } from "../model/config.js";

import ropeWGSL from "../shaders/rope.wgsl";
import softmaxWGSL from "../shaders/softmax.wgsl";
import attentionWGSL from "../shaders/attention.wgsl";

/**
 * Multi-Head Attention with support for GQA (Grouped Query Attention).
 *
 * Components:
 * - Q/K/V projections via BitLinear
 * - RoPE (Rotary Position Embeddings)
 * - Scaled dot-product attention with causal mask
 * - Output projection via BitLinear
 * - KV-cache for autoregressive generation
 */
export class Attention {
  private device: GPUDevice;
  private pipelines: PipelineManager;
  private pool: BufferPool;
  private config: ModelConfig;
  private hDim: number;

  private qProj: BitLinear;
  private kProj: BitLinear;
  private vProj: BitLinear;
  private oProj: BitLinear;

  // Pre-allocated uniform buffers for N=1 decode (dynamic — updated via writeBuffer)
  private decodeRopeQUniform?: GPUBuffer;
  private decodeRopeKUniform?: GPUBuffer;
  private decodeScoresUniform?: GPUBuffer;
  private decodeSoftmaxUniform?: GPUBuffer;
  private decodeAttnVUniform?: GPUBuffer;

  // Pre-allocated uniform buffers for N>1 prefill (dynamic — updated via writeBuffer)
  private prefillRopeQUniform?: GPUBuffer;
  private prefillRopeKUniform?: GPUBuffer;
  private prefillScoresUniform?: GPUBuffer;
  private prefillSoftmaxUniform?: GPUBuffer;
  private prefillAttnVUniform?: GPUBuffer;

  // Pre-allocated score/attnWeight buffers for N=1 decode (sized to maxSeqLen)
  private decodeScoresBuf?: GPUBuffer;
  private decodeAttnWeightsBuf?: GPUBuffer;

  // Bind group cache for N=1 decode
  private bgCache: BindGroupCache = createBGCache();

  constructor(
    device: GPUDevice,
    pipelines: PipelineManager,
    pool: BufferPool,
    config: ModelConfig,
    qProj: BitLinear,
    kProj: BitLinear,
    vProj: BitLinear,
    oProj: BitLinear
  ) {
    this.device = device;
    this.pipelines = pipelines;
    this.pool = pool;
    this.config = config;
    this.hDim = headDim(config);
    this.qProj = qProj;
    this.kProj = kProj;
    this.vProj = vProj;
    this.oProj = oProj;
  }

  /** Pre-allocate uniform buffers for N=1 decode (updated via writeBuffer each token). */
  initDecodeUniforms(maxSeqLen: number): void {
    const mkBuf = (size: number) =>
      this.device.createBuffer({
        size,
        usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
      });
    this.decodeRopeQUniform = mkBuf(20);
    this.decodeRopeKUniform = mkBuf(20);
    this.decodeScoresUniform = mkBuf(24);
    this.decodeSoftmaxUniform = mkBuf(8);
    this.decodeAttnVUniform = mkBuf(20);

    // Prefill uniforms (reused via writeBuffer for N>1)
    this.prefillRopeQUniform = mkBuf(20);
    this.prefillRopeKUniform = mkBuf(20);
    this.prefillScoresUniform = mkBuf(24);
    this.prefillSoftmaxUniform = mkBuf(8);
    this.prefillAttnVUniform = mkBuf(20);

    // Pre-allocate score/attnWeight buffers at maxSeqLen so buffer identity
    // is stable across decode tokens (totalSeq grows but buffer stays the same)
    const maxScoresSize = this.config.numAttentionHeads * maxSeqLen * 4;
    this.decodeScoresBuf = this.device.createBuffer({
      size: maxScoresSize,
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC,
    });
    this.decodeAttnWeightsBuf = this.device.createBuffer({
      size: maxScoresSize,
      usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC,
    });

    this.qProj.initDecodeUniforms();
    this.kProj.initDecodeUniforms();
    this.vProj.initDecodeUniforms();
    this.oProj.initDecodeUniforms();
  }

  /**
   * Forward pass.
   * @param input [N, hidden] f32
   * @param N sequence length
   * @param kvCache KV cache for autoregressive decoding
   * @param encoder GPU command encoder
   * @returns output [N, hidden] f32
   */
  forward(
    input: GPUBuffer,
    N: number,
    kvCache: KVCache,
    encoder: GPUCommandEncoder
  ): GPUBuffer {
    const { numAttentionHeads, numKeyValueHeads, hiddenSize } = this.config;

    // Q/K/V projections via BitLinear
    const qBuf = this.qProj.forward(input, N, encoder);
    const kBuf = this.kProj.forward(input, N, encoder);
    const vBuf = this.vProj.forward(input, N, encoder);

    // Apply RoPE to Q and K
    const qRoped = this.applyRoPE(
      encoder,
      qBuf,
      N,
      numAttentionHeads,
      kvCache.seqLen,
      N === 1 ? this.decodeRopeQUniform : this.prefillRopeQUniform,
      "ropeQ"
    );
    const kRoped = this.applyRoPE(
      encoder,
      kBuf,
      N,
      numKeyValueHeads,
      kvCache.seqLen,
      N === 1 ? this.decodeRopeKUniform : this.prefillRopeKUniform,
      "ropeK"
    );

    this.pool.release(qBuf);
    this.pool.release(kBuf);

    // Update KV cache
    this.appendToCache(encoder, kRoped, vBuf, kvCache, N);
    this.pool.release(kRoped);
    this.pool.release(vBuf);

    const totalSeq = kvCache.seqLen + N;

    // Compute attention scores: Q @ K^T * scale
    const scores = this.computeScores(
      encoder,
      qRoped,
      kvCache.key,
      N,
      totalSeq,
      N === 1 ? this.decodeScoresUniform : this.prefillScoresUniform,
      N === 1 ? this.decodeScoresBuf : undefined
    );
    this.pool.release(qRoped);

    // Softmax over scores
    const attnWeights = this.applySoftmax(
      encoder,
      scores,
      numAttentionHeads * N,
      totalSeq,
      N === 1 ? this.decodeSoftmaxUniform : this.prefillSoftmaxUniform,
      N === 1 ? this.decodeAttnWeightsBuf : undefined
    );
    if (N !== 1) this.pool.release(scores);

    // Attention output: weights @ V
    const attnOutput = this.computeAttnV(
      encoder,
      attnWeights,
      kvCache.value,
      N,
      totalSeq,
      N === 1 ? this.decodeAttnVUniform : this.prefillAttnVUniform
    );
    if (N !== 1) this.pool.release(attnWeights);

    // Output projection via BitLinear
    const output = this.oProj.forward(attnOutput, N, encoder);
    this.pool.release(attnOutput);

    return output;
  }

  private applyRoPE(
    encoder: GPUCommandEncoder,
    input: GPUBuffer,
    N: number,
    numHeads: number,
    posOffset: number,
    uniformBuf?: GPUBuffer,
    bgId?: string
  ): GPUBuffer {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "rope",
      ropeWGSL
    );

    const outputSize = N * numHeads * this.hDim * 4;
    const output = this.pool.acquire(
      outputSize,
      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
    );

    const paramsData = new ArrayBuffer(20);
    const paramsView = new DataView(paramsData);
    paramsView.setUint32(0, N, true);
    paramsView.setUint32(4, numHeads, true);
    paramsView.setUint32(8, this.hDim, true);
    paramsView.setUint32(12, posOffset, true);
    paramsView.setFloat32(16, this.config.ropeTheta, true);

    let paramsBuffer: GPUBuffer;
    if (uniformBuf) {
      this.device.queue.writeBuffer(uniformBuf, 0, new Uint8Array(paramsData));
      paramsBuffer = uniformBuf;
    } else {
      paramsBuffer = this.createUniform(paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: input } },
      { binding: 1, resource: { buffer: output } },
      { binding: 2, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1 && bgId
      ? cachedBG(this.bgCache, this.device, bgId, bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const totalPairs = N * numHeads * (this.hDim / 2);
    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(Math.ceil(totalPairs / 256));
    pass.end();

    return output;
  }

  private appendToCache(
    encoder: GPUCommandEncoder,
    k: GPUBuffer,
    v: GPUBuffer,
    cache: KVCache,
    N: number
  ): void {
    const kvSize =
      N * this.config.numKeyValueHeads * this.hDim * 4;
    const offset = cache.seqLen * this.config.numKeyValueHeads * this.hDim * 4;

    encoder.copyBufferToBuffer(k, 0, cache.key, offset, kvSize);
    encoder.copyBufferToBuffer(v, 0, cache.value, offset, kvSize);
  }

  private computeScores(
    encoder: GPUCommandEncoder,
    Q: GPUBuffer,
    K: GPUBuffer,
    N: number,
    S: number,
    uniformBuf?: GPUBuffer,
    preAllocated?: GPUBuffer
  ): GPUBuffer {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "attention_scores",
      attentionWGSL,
      "compute_scores"
    );

    const { numAttentionHeads, numKeyValueHeads } = this.config;
    const scores = preAllocated ?? this.pool.acquire(
      numAttentionHeads * N * S * 4,
      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
    );

    const paramsData = new ArrayBuffer(24);
    const v = new DataView(paramsData);
    v.setUint32(0, N, true);
    v.setUint32(4, S, true);
    v.setUint32(8, numAttentionHeads, true);
    v.setUint32(12, numKeyValueHeads, true);
    v.setUint32(16, this.hDim, true);
    v.setFloat32(20, 1.0 / Math.sqrt(this.hDim), true);

    let paramsBuffer: GPUBuffer;
    if (uniformBuf) {
      this.device.queue.writeBuffer(uniformBuf, 0, new Uint8Array(paramsData));
      paramsBuffer = uniformBuf;
    } else {
      paramsBuffer = this.createUniform(paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: Q } },
      { binding: 1, resource: { buffer: K } },
      { binding: 2, resource: { buffer: scores } },
      { binding: 3, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1
      ? cachedBG(this.bgCache, this.device, "scores", bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(
      Math.ceil(N / 16),
      Math.ceil(S / 16),
      numAttentionHeads
    );
    pass.end();

    return scores;
  }

  private applySoftmax(
    encoder: GPUCommandEncoder,
    input: GPUBuffer,
    N: number,
    D: number,
    uniformBuf?: GPUBuffer,
    preAllocated?: GPUBuffer
  ): GPUBuffer {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "softmax",
      softmaxWGSL
    );

    const output = preAllocated ?? this.pool.acquire(
      N * D * 4,
      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
    );

    const paramsData = new ArrayBuffer(8);
    const v = new DataView(paramsData);
    v.setUint32(0, N, true);
    v.setUint32(4, D, true);

    let paramsBuffer: GPUBuffer;
    if (uniformBuf) {
      this.device.queue.writeBuffer(uniformBuf, 0, new Uint8Array(paramsData));
      paramsBuffer = uniformBuf;
    } else {
      paramsBuffer = this.createUniform(paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: input } },
      { binding: 1, resource: { buffer: output } },
      { binding: 2, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1
      ? cachedBG(this.bgCache, this.device, "softmax", bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(N);
    pass.end();

    return output;
  }

  private computeAttnV(
    encoder: GPUCommandEncoder,
    attn: GPUBuffer,
    V: GPUBuffer,
    N: number,
    S: number,
    uniformBuf?: GPUBuffer
  ): GPUBuffer {
    const { pipeline, bindGroupLayout } = this.pipelines.getOrCreate(
      "attn_v",
      attentionWGSL,
      "attn_v"
    );

    const { numAttentionHeads, numKeyValueHeads } = this.config;
    const outputSize = N * numAttentionHeads * this.hDim * 4;
    const output = this.pool.acquire(
      outputSize,
      GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC
    );

    const paramsData = new ArrayBuffer(20);
    const v = new DataView(paramsData);
    v.setUint32(0, N, true);
    v.setUint32(4, S, true);
    v.setUint32(8, numAttentionHeads, true);
    v.setUint32(12, numKeyValueHeads, true);
    v.setUint32(16, this.hDim, true);

    let paramsBuffer: GPUBuffer;
    if (uniformBuf) {
      this.device.queue.writeBuffer(uniformBuf, 0, new Uint8Array(paramsData));
      paramsBuffer = uniformBuf;
    } else {
      paramsBuffer = this.createUniform(paramsData);
    }

    const entries: GPUBindGroupEntry[] = [
      { binding: 0, resource: { buffer: attn } },
      { binding: 1, resource: { buffer: V } },
      { binding: 2, resource: { buffer: output } },
      { binding: 3, resource: { buffer: paramsBuffer } },
    ];
    const bindGroup = N === 1
      ? cachedBG(this.bgCache, this.device, "attnV", bindGroupLayout, entries)
      : this.device.createBindGroup({ layout: bindGroupLayout, entries });

    const total = N * numAttentionHeads * this.hDim;
    const pass = encoder.beginComputePass();
    pass.setPipeline(pipeline);
    pass.setBindGroup(0, bindGroup);
    pass.dispatchWorkgroups(Math.ceil(total / 256));
    pass.end();

    return output;
  }

  /** Clear the bind group cache (call on KV cache reset). */
  clearBGCache(): void {
    clearBGCache(this.bgCache);
    this.qProj.clearBGCache();
    this.kProj.clearBGCache();
    this.vProj.clearBGCache();
    this.oProj.clearBGCache();
  }

  /** Destroy pre-allocated attention buffers. */
  destroyPreAllocated(): void {
    this.decodeScoresBuf?.destroy();
    this.decodeAttnWeightsBuf?.destroy();
    this.decodeScoresBuf = undefined;
    this.decodeAttnWeightsBuf = undefined;
  }

  private createUniform(data: ArrayBuffer): GPUBuffer {
    const size = Math.max(Math.ceil(data.byteLength / 4) * 4, 4);
    const buffer = this.device.createBuffer({
      size,
      usage: GPUBufferUsage.UNIFORM | GPUBufferUsage.COPY_DST,
      mappedAtCreation: true,
    });
    new Uint8Array(buffer.getMappedRange()).set(new Uint8Array(data));
    buffer.unmap();
    return buffer;
  }
}

/** Allocate a fresh KV cache for a given config and max sequence length */
export function createKVCache(
  device: GPUDevice,
  config: ModelConfig,
  maxSeqLen: number
): KVCache {
  const kvSize =
    maxSeqLen * config.numKeyValueHeads * headDim(config) * 4;
  const key = device.createBuffer({
    size: kvSize,
    usage:
      GPUBufferUsage.STORAGE |
      GPUBufferUsage.COPY_DST |
      GPUBufferUsage.COPY_SRC,
  });
  const value = device.createBuffer({
    size: kvSize,
    usage:
      GPUBufferUsage.STORAGE |
      GPUBufferUsage.COPY_DST |
      GPUBufferUsage.COPY_SRC,
  });
  return { key, value, seqLen: 0, maxSeqLen };
}

File: packages/rust/crates/oxbitnet/src/nn/attention.rs
use std::sync::Arc;

use wgpu::BufferUsages;

use crate::gpu::buffer_pool::{BufferPool, GpuBuf};
use crate::gpu::pipeline::PipelineManager;
use crate::model::config::ModelConfig;
use crate::nn::bg_cache::BgCache;
use crate::nn::bitlinear::{buf_entry, create_uniform_raw, BitLinear};

const ROPE_WGSL: &str = include_str!("../shaders/rope.wgsl");
const SOFTMAX_WGSL: &str = include_str!("../shaders/softmax.wgsl");
const ATTENTION_WGSL: &str = include_str!("../shaders/attention.wgsl");

/// KV cache for autoregressive generation.
pub struct KvCache {
    pub key: wgpu::Buffer,
    pub value: wgpu::Buffer,
    pub seq_len: usize,
    pub max_seq_len: usize,
}

pub fn create_kv_cache(
    device: &wgpu::Device,
    config: &ModelConfig,
    max_seq_len: usize,
) -> KvCache {
    let kv_size =
        (max_seq_len * config.num_key_value_heads * config.head_dim() * 4) as u64;
    let key = device.create_buffer(&wgpu::BufferDescriptor {
        label: Some("kv_key"),
        size: kv_size,
        usage: BufferUsages::STORAGE | BufferUsages::COPY_DST | BufferUsages::COPY_SRC,
        mapped_at_creation: false,
    });
    let value = device.create_buffer(&wgpu::BufferDescriptor {
        label: Some("kv_value"),
        size: kv_size,
        usage: BufferUsages::STORAGE | BufferUsages::COPY_DST | BufferUsages::COPY_SRC,
        mapped_at_creation: false,
    });
    KvCache {
        key,
        value,
        seq_len: 0,
        max_seq_len,
    }
}

/// Multi-Head Attention with GQA support.
pub struct Attention {
    device: Arc<wgpu::Device>,
    config: ModelConfig,
    h_dim: usize,
    pub(crate) q_proj: BitLinear,
    pub(crate) k_proj: BitLinear,
    pub(crate) v_proj: BitLinear,
    pub(crate) o_proj: BitLinear,
    bg_cache: BgCache,
}

impl Attention {
    pub fn new(
        device: Arc<wgpu::Device>,
        config: ModelConfig,
        q_proj: BitLinear,
        k_proj: BitLinear,
        v_proj: BitLinear,
        o_proj: BitLinear,
    ) -> Self {
        let h_dim = config.head_dim();
        Self {
            device,
            config,
            h_dim,
            q_proj,
            k_proj,
            v_proj,
            o_proj,
            bg_cache: BgCache::new(),
        }
    }

    /// Forward pass: input [N, hidden] → output [N, hidden]
    pub fn forward(
        &mut self,
        input: &GpuBuf,
        n: usize,
        kv_cache: &mut KvCache,
        encoder: &mut wgpu::CommandEncoder,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        let num_heads = self.config.num_attention_heads;
        let num_kv_heads = self.config.num_key_value_heads;

        // Q/K/V projections
        let q_buf = self.q_proj.forward(input, n, encoder, pipelines, pool);
        let k_buf = self.k_proj.forward(input, n, encoder, pipelines, pool);
        let v_buf = self.v_proj.forward(input, n, encoder, pipelines, pool);

        // RoPE
        let q_roped = self.apply_rope(encoder, &q_buf, n, num_heads, kv_cache.seq_len, pipelines, pool);
        let k_roped = self.apply_rope(encoder, &k_buf, n, num_kv_heads, kv_cache.seq_len, pipelines, pool);

        // Update KV cache
        self.append_to_cache(encoder, &k_roped, &v_buf, kv_cache, n);

        let total_seq = kv_cache.seq_len + n;

        // Attention scores: Q @ K^T * scale
        let scores = self.compute_scores(encoder, &q_roped, &kv_cache.key, n, total_seq, pipelines, pool);

        // Softmax
        let attn_weights = self.apply_softmax(encoder, &scores, num_heads * n, total_seq, pipelines, pool);

        // Attention output: weights @ V
        let attn_output = self.compute_attn_v(encoder, &attn_weights, &kv_cache.value, n, total_seq, pipelines, pool);

        // Output projection
        let output = self.o_proj.forward(&attn_output, n, encoder, pipelines, pool);

        output
    }

    fn apply_rope(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        input: &wgpu::Buffer,
        n: usize,
        num_heads: usize,
        pos_offset: usize,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        let entry = pipelines.get_or_create_default("rope", ROPE_WGSL);

        let output_size = (n * num_heads * self.h_dim * 4) as u64;
        let output = pool.acquire(output_size, BufferUsages::STORAGE | BufferUsages::COPY_SRC);

        let params_data = [
            (n as u32).to_le_bytes(),
            (num_heads as u32).to_le_bytes(),
            (self.h_dim as u32).to_le_bytes(),
            (pos_offset as u32).to_le_bytes(),
            self.config.rope_theta.to_le_bytes(),
        ]
        .concat();
        let params = create_uniform_raw(&self.device, &params_data);

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("rope"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, input),
                buf_entry(1, &output),
                buf_entry(2, &params),
            ],
        });

        let total_pairs = (n * num_heads * (self.h_dim / 2)) as u32;
        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups((total_pairs + 255) / 256, 1, 1);

        output
    }

    fn append_to_cache(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        k: &wgpu::Buffer,
        v: &wgpu::Buffer,
        cache: &KvCache,
        n: usize,
    ) {
        let kv_size = (n * self.config.num_key_value_heads * self.h_dim * 4) as u64;
        let offset = (cache.seq_len * self.config.num_key_value_heads * self.h_dim * 4) as u64;

        encoder.copy_buffer_to_buffer(k, 0, &cache.key, offset, kv_size);
        encoder.copy_buffer_to_buffer(v, 0, &cache.value, offset, kv_size);
    }

    fn compute_scores(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        q: &wgpu::Buffer,
        k: &wgpu::Buffer,
        n: usize,
        s: usize,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        let entry = pipelines.get_or_create("attention_scores", ATTENTION_WGSL, "compute_scores", None);
        let num_heads = self.config.num_attention_heads;
        let num_kv_heads = self.config.num_key_value_heads;

        let scores = pool.acquire(
            (num_heads * n * s * 4) as u64,
            BufferUsages::STORAGE | BufferUsages::COPY_SRC,
        );

        let scale = 1.0 / (self.h_dim as f32).sqrt();
        let params_data = [
            (n as u32).to_le_bytes(),
            (s as u32).to_le_bytes(),
            (num_heads as u32).to_le_bytes(),
            (num_kv_heads as u32).to_le_bytes(),
            (self.h_dim as u32).to_le_bytes(),
            scale.to_le_bytes(),
        ]
        .concat();
        let params = create_uniform_raw(&self.device, &params_data);

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("attn_scores"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, q),
                buf_entry(1, k),
                buf_entry(2, &scores),
                buf_entry(3, &params),
            ],
        });

        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups(
            ((n + 15) / 16) as u32,
            ((s + 15) / 16) as u32,
            num_heads as u32,
        );

        scores
    }

    fn apply_softmax(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        input: &wgpu::Buffer,
        n: usize,
        d: usize,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        let entry = pipelines.get_or_create_default("softmax", SOFTMAX_WGSL);

        let output = pool.acquire(
            (n * d * 4) as u64,
            BufferUsages::STORAGE | BufferUsages::COPY_SRC,
        );

        let params_data = [
            (n as u32).to_le_bytes(),
            (d as u32).to_le_bytes(),
        ]
        .concat();
        let params = create_uniform_raw(&self.device, &params_data);

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("softmax"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, input),
                buf_entry(1, &output),
                buf_entry(2, &params),
            ],
        });

        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups(n as u32, 1, 1);

        output
    }

    fn compute_attn_v(
        &self,
        encoder: &mut wgpu::CommandEncoder,
        attn: &wgpu::Buffer,
        v: &wgpu::Buffer,
        n: usize,
        s: usize,
        pipelines: &mut PipelineManager,
        pool: &BufferPool,
    ) -> GpuBuf {
        let entry = pipelines.get_or_create("attn_v", ATTENTION_WGSL, "attn_v", None);
        let num_heads = self.config.num_attention_heads;
        let num_kv_heads = self.config.num_key_value_heads;

        let output_size = (n * num_heads * self.h_dim * 4) as u64;
        let output = pool.acquire(output_size, BufferUsages::STORAGE | BufferUsages::COPY_SRC);

        let params_data = [
            (n as u32).to_le_bytes(),
            (s as u32).to_le_bytes(),
            (num_heads as u32).to_le_bytes(),
            (num_kv_heads as u32).to_le_bytes(),
            (self.h_dim as u32).to_le_bytes(),
        ]
        .concat();
        let params = create_uniform_raw(&self.device, &params_data);

        let bg = self.device.create_bind_group(&wgpu::BindGroupDescriptor {
            label: Some("attn_v"),
            layout: &entry.bind_group_layout,
            entries: &[
                buf_entry(0, attn),
                buf_entry(1, v),
                buf_entry(2, &output),
                buf_entry(3, &params),
            ],
        });

        let total = (n * num_heads * self.h_dim) as u32;
        let mut pass = encoder.begin_compute_pass(&Default::default());
        pass.set_pipeline(&entry.pipeline);
        pass.set_bind_group(0, Some(&bg), &[]);
        pass.dispatch_workgroups((total + 255) / 256, 1, 1);

        output
    }

    pub fn clear_bg_cache(&mut self) {
        self.bg_cache.clear();
        self.q_proj.clear_bg_cache();
        self.k_proj.clear_bg_cache();
        self.v_proj.clear_bg_cache();
        self.o_proj.clear_bg_cache();
    }
}
